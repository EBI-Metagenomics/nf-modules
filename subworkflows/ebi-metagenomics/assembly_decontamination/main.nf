include { DECONTAMINATE_CONTIGS as HUMAN_DECONTAMINATE_CONTIGS } from '../decontaminate_contigs/main'
include { DECONTAMINATE_CONTIGS as PHIX_DECONTAMINATE_CONTIGS  } from '../decontaminate_contigs/main'
include { DECONTAMINATE_CONTIGS as HOST_DECONTAMINATE_CONTIGS  } from '../decontaminate_contigs/main'

workflow ASSEMBLY_DECONTAMINATION {
    /*
    * Microbiome Informatics metagenomics assembly decontamination subworkflow
    *
    * Performs sequential decontamination of assembled contigs against the human genome, PhiX and a genome
    * specified by the user (usually the sample host). It is designed to remove any contigs that come from any
    * of those references.
    *
    * Rationale:
    * 1. PhiX sequences - commonly used as spike-in controls in Illumina sequencing
    * 2. Human sequences - to remove human contamination from samples
    * 3. Host/contaminant sequences - custom contaminants specific to the experimental design
    *
    * Each decontamination step is optional and controlled by metadata parameters, allowing
    * flexible configuration based on sample requirements.
    *
    * Global parameters:
    * - params.reference_genomes_folder: Path to directory containing reference genome files
    *
    * Parameters (via meta map):
    *
    * - meta.phix_reference: Filename of PhiX reference genome (optional)
    *   - If null, PhiX decontamination is skipped
    *   - File must exist in params.reference_genomes_folder
    *
    * - meta.human_reference: Filename of human reference genome (optional)
    *   - If null, human decontamination is skipped
    *   - File must exist in params.reference_genomes_folder
    *
    * - meta.contaminant_reference: Filename of custom contaminant reference (optional)
    *   - If null, host/contaminant decontamination is skipped
    *   - File must exist in params.reference_genomes_folder
    *
    * Input:
    * - assembly: Channel [ val(meta), path(assembly_fasta) ]
    *
    * Outputs:
    * - cleaned_contigs: Channel [ val(meta), path(cleaned_assembly_fasta) ]
    * - versions: Channel containing software version information
    *
    */

    take:
    assembly               // [ val(meta), path(assembly_fasta) ]

    main:
    ch_versions = Channel.empty()

    /***************************************************************************/
    /* Perform decontamination from PhiX sequences if requested                */
    /***************************************************************************/
    assembly.branch { meta, _contigs ->
            run_decontamination: meta.phix_reference != null
            skip_decontamination: meta.phix_reference == null
        }
        .set { phix_subdivided_assemblies }

    phix_subdivided_assemblies.run_decontamination
        .map { meta, contigs ->
            [ [meta, contigs], file( "${params.reference_genomes_folder}/${meta.phix_reference}", checkIfExists: true )  ]
        }
        .set { ch_phix_decontamination_input }

    PHIX_DECONTAMINATE_CONTIGS(ch_phix_decontamination_input)
    ch_versions = ch_versions.mix(PHIX_DECONTAMINATE_CONTIGS.out.versions)

    phix_cleaned_contigs = phix_subdivided_assemblies.skip_decontamination.mix(
        PHIX_DECONTAMINATE_CONTIGS.out.cleaned_contigs
    )

    /***************************************************************************/
    /* Perform decontamination from human sequences if requested               */
    /***************************************************************************/
    phix_cleaned_contigs
        .branch { meta, _contigs ->
            run_decontamination: meta.human_reference != null
            skip_decontamination: meta.human_reference == null
        }
        .set { human_subdivided_assemblies }

    human_subdivided_assemblies.run_decontamination
        .map { meta, contigs ->
            [ [meta, contigs], file( "${params.reference_genomes_folder}/${meta.human_reference}", checkIfExists: true ) ]
        }
        .set { ch_human_decontamination_input }

    HUMAN_DECONTAMINATE_CONTIGS(ch_human_decontamination_input)
    ch_versions = ch_versions.mix(HUMAN_DECONTAMINATE_CONTIGS.out.versions)

    human_cleaned_contigs = human_subdivided_assemblies.skip_decontamination.mix(
        HUMAN_DECONTAMINATE_CONTIGS.out.cleaned_contigs
    )

    /***************************************************************************/
    /* Perform decontamination from arbitrary contaminant sequences            */
    /***************************************************************************/
    human_cleaned_contigs
        .branch { meta, _contigs ->
            run_decontamination: meta.contaminant_reference != null
            skip_decontamination: meta.contaminant_reference == null
        }
        .set { subdivided_assemblies }

    subdivided_assemblies.run_decontamination
        .map { meta, contigs ->
            [ [meta, contigs], file( "${params.reference_genomes_folder}/${meta.contaminant_reference}", checkIfExists: true ) ]
        }
        .set { ch_decontamination_input }

    HOST_DECONTAMINATE_CONTIGS(ch_decontamination_input)
    ch_versions = ch_versions.mix(HOST_DECONTAMINATE_CONTIGS.out.versions)

    cleaned_contigs = subdivided_assemblies.skip_decontamination.mix(
        HOST_DECONTAMINATE_CONTIGS.out.cleaned_contigs
    )

    emit:
    cleaned_contigs = cleaned_contigs
    versions        = ch_versions
}
