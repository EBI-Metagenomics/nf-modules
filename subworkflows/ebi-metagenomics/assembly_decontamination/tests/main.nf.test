nextflow_workflow {
    
    name "Test subworkflow ASSEMBLY_DECONTAMINATION"
    script "../main.nf"
    config "./nextflow.config"
    workflow "ASSEMBLY_DECONTAMINATION"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/assembly_decontamination"
    tag "blast/blastn"
    tag "seqkit/grep"

    test("Assembly host decontamination") {
        when {
            params {
                blast_reference_genomes_folder    = "${moduleDir}/tests/data"
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ], // meta map
                    file("${moduleDir}/tests/data/test.fa", checkIfExists: true)
                ])
                input[1] = "contamination.fa"
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.cleaned_contigs).match() }
            )
        }
    }
}
