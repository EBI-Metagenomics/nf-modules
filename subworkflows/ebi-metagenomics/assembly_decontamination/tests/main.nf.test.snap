{
    "assembly_decontamination - multiple samples": {
        "content": [
            [
                [
                    {
                        "id": "sample1",
                        "phix_reference": "phyx",
                        "human_reference": "human",
                        "contaminant_reference": null
                    },
                    "sample1.fa.gz:md5,a9e09e2aa32fa94543bab203db8c65e2"
                ],
                [
                    {
                        "id": "sample2",
                        "phix_reference": null,
                        "human_reference": null,
                        "contaminant_reference": "contamination"
                    },
                    "sample2.fa.gz:md5,e94629f8dec8ddd183beaae566c95c3f"
                ]
            ],
            [
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-08-05T16:11:25.408034"
    },
    "assembly_decontamination - phix only": {
        "content": [
            [
                [
                    {
                        "id": "test_sample_phix",
                        "phix_reference": "phyx",
                        "human_reference": null,
                        "contaminant_reference": null
                    },
                    "test_sample_phix.fa.gz:md5,99855030a1f86513a54dfbb7c4dfcb33"
                ]
            ],
            [
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-08-05T16:11:08.355247"
    },
    "assembly_decontamination - host only": {
        "content": [
            [
                [
                    {
                        "id": "test_sample_host",
                        "phix_reference": null,
                        "human_reference": null,
                        "contaminant_reference": "contamination"
                    },
                    "test_sample_host.fa.gz:md5,e94629f8dec8ddd183beaae566c95c3f"
                ]
            ],
            [
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-08-05T16:11:17.964707"
    },
    "assembly_decontamination - all references": {
        "content": [
            [
                [
                    {
                        "id": "test_sample_all",
                        "phix_reference": "phyx",
                        "human_reference": "human",
                        "contaminant_reference": "contamination"
                    },
                    "test_sample_all.fa.gz:md5,25aa49239f327c6704ad4590eb1698c9"
                ]
            ],
            [
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HOST_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:PHIX_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-08-05T16:11:02.969589"
    },
    "assembly_decontamination - human only": {
        "content": [
            [
                [
                    {
                        "id": "test_sample_human",
                        "phix_reference": null,
                        "human_reference": "human",
                        "contaminant_reference": null
                    },
                    "test_sample_human.fa.gz:md5,34ac8d3b788517e6aae1c5e7c63c720a"
                ]
            ],
            [
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:FILTERPAF": {
                        "gawk": "5.3.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:SEQKIT_GREP": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "ASSEMBLY_DECONTAMINATION:HUMAN_DECONTAMINATE_CONTIGS:MINIMAP2_ALIGN": {
                        "minimap2": "2.29-r1283",
                        "samtools": 1.21
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-08-05T16:11:13.132533"
    }
}